inverse/private/ft_inv.m

```diff
@@ -0,0 +1,179 @@
+interactive = istrue(ft_getopt(varargin, 'interactive', false));
+
+[m, n] = size(A);
+if m==0 || n==0
+  Ainv = A';
+  return;
+end
+
+if n < m
+  % recurse into ft_inv on the transposed input
+  Ainv = ft_inv(A',varargin{:})';
+  return;
+end
+
+% first deal with the lambda argument and regularize
+if ischar(lambda) && lambda(end)=='%'
+  ratio  = sscanf(lambda, '%f%%');
+  ratio  = ratio/100;
+  lambda = ratio * trace(A)./m;
+end
+
+if lambda~=0 && m~=n
+  ft_error('with lambda regularization, the input matrix should be square');
+end
+
+if m==n && lambda>0
+  % regularize
+  A = A + eye(m).*lambda;
+elseif m~=n && lambda~=0
+  ft_error('when lambda larger than zero, the input matrix must be square');
+elseif m==n && lambda<0
+  ft_error('a negative value for lambda is not allowed');
+end
+
+% perform an svd
+[U,S,V] = svd(A,0);
+if m > 1
+  s = diag(S);
+elseif m == 1
+  s = S(1);
+else
+  s = 0;
+end
+
+if isempty(kappa) && isempty(tolerance)
+  % check whether the interacive mode is requested
+  if interactive
+    figure; plot(1:m, log10(s),'o-');
+    kappa = input('Specify the number of dimensions at which the eigenvalue spectrum will be truncated: ');
+    if isempty(kappa)
+      kappa = m;
+    end
+    usekappa     = true;
+    usetolerance = false;
+  else
+    % The default tolerance here is higher than the one used in MATLAB's
+    % pinv, with a factor of 10. This was done by Robert Oostenveld in 2004,
+    % to avoid numerical tolerance issues in the MATLAB version that was
+    % used back then
+    tolerance = 10 * max(m,n) * eps;
+    kappa     = sum(s./s(1) > tolerance);
+    
+    usekappa     = false;
+    usetolerance = true;
+  end
+elseif ~isempty(kappa) &&  isempty(tolerance)
+  usekappa     = true;
+  usetolerance = false;
+elseif  isempty(kappa) && ~isempty(tolerance)
+  usekappa     = false;
+  usetolerance = true; % needed for winsorization
+  kappa = sum(s./s(1) > tolerance);
+elseif ~isempty(kappa) && ~isempty(tolerance)
+  ft_warning('both a kappa value and tolerance value are defined, using the kappa value for truncation');
+  usekappa     = true;
+  usetolerance = false;
+end
+
+% do a sanity check on the estimated kappa
+if kappa<=0 || kappa>m
+  ft_warning('The kappa value falls outside the supported range, returning an all zeros matrix');
+  Ainv = zeros(size(A'),class(A));
+  return;
+end
+
+switch method
+  case 'vanilla'
+    Ainv = inv(A);
+  case 'svd'
+    S    = diag(ones(kappa,1)./s(1:kappa));
+    Ainv = V(:,1:kappa)*S*U(:,1:kappa)';
+  case 'moorepenrose'
+    Ainv = A'/(A*A');
+  case 'winsorize'
+    % replace all singular values < the kappa'th singular by the kappa'th
+    % singular value
+    if usekappa
+      s = max(s,s(kappa));
+    elseif usetolerance
+      s(s./s(1) < tolerance) = tolerance;
+    end
+    S    = diag(1./s);
+    Ainv = V*S*U';
+  otherwise
+    ft_error('Unsupported method for matrix inversion');
+end
+
+if feedback
+  figure; hold on
```

Comment:
It would be useful to have an option to pass a handle to a figure as we do e.g.  in ft_average_sense. That way I can do the plot in SPM graphics window.

Reply:
does this still work in newer matlab versions? In the past we used numbers for handles and that broke down when matlab upgraded their graphics handling

Reply:
I don't know. I haven't had any problems with that but I still use 2014/2015

Reply:
Guillaume says handles are objects in the newer version so as long as you just handle them as some variable and don't assume they are a number, there shouldn't be a problem.